core/primitives-core/src/profile.rs

```diff
@@ -13,6 +16,10 @@ pub struct ProfileData {
     data: Rc<DataArray>,
 }
 
+/// Result of a ProfileData stat
+#[derive(Clone, BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct ProfileDataResult(pub Vec<u64>);
```

Comment:
This type is going to become part of the state, so I think it makes sense to spend some effort designing it with proper encapsulation and such. Just exposing the raw `Vec<u64>` feels like it might make changing this later complicated. 

I see the following issues here:

* The name is `??Result`, but this isn't an alias to `std::result::Result`. It's better to resolve Result suffix for result *enums*. Here, I think the better name would be `GasProfile`. 
* The field is public, which makes it inconvenient for clients to use this (as the data is in a very raw form), and for the implementation to evolve (ie, if we decide to use sparse, rather than dense, storage here, we'd have to change every call-site)
* Derived serialization and deserialization representations don't take evolvability into the account. For example, if we add new `ActionCosts` variant, then all `ExtCosts` indices will shift by one, which will change the meaning of data already stored in the database. 

I think we should do the following:

* move this type into a dedeicated module (it *can* be a well-isolated value type, so it makes sense to enforce it)
* rename it to `GasProfile`
* make the impl private and provide accessors. I think we need two assessors here: an `impl fmt::Display for GasProfile`, so that it's easy to debug stuff in the repo. Additionally, we need something like `fn nonzero_costs(&self) -> BTreeMap<Cost, u64>`, where `Cost` is either an `ActionCost` or an `ExtCost`. Logically, the data here is a mapping from cost metrics to their values, and we should present it to the clients as such. Probably the best interface here would look like this:

```rust
impl Cost {
    const ALL: &'static [COST] = &[...];
}

impl Index<Cost> for GasProfile {
    type Output = u64;
}
```
* do *not* derive `Serialize` -- for JSON RPC, I think we want to serialize this just as a map from string keys to numeric values, and that should be doable in the rpc crate using the ALL/Index API. If `serialize` is required here, because the RPC just want to dump ExecutionOutcome to JSON as is, we should implement `Serialize` manually such that the representation is a map with string keys. 
* do implement `BorshSerialize`, and make sure that its forward compatible with addition of new acction and ext costs. In parcitular, the current format where we first store all actions and then all costs is *not* forward compatible, as adding any action invalidates all ext costs.

Reply:
> This type is going to become part of the state

I was under the impression that it was not going to be part of the state. Could you point me to where it is saved to the state @matklad ?

Reply:
We are stuffing this into `ExecutionMetadata`, which is a part of `ExecutionOutcome`, which is stored in the `ColTransactionResult` column. I might be using terminology incorrectly: it is a part of the state we store in rocksdb, but it isn't the part of the state which affects merkle root. That is, it's not a part of `PartialExecutionOutcome`. 

Do we have terms for:

* stuff we store in rocks and which doesn't affect merkle tree
* stuff we store in merkle tree?

Reply:
> stuff we store in merkle tree?

That is the blockchain state.

> stuff we store in rocks and which doesn't affect merkle tree

I don't think we have a good term for that.

Reply:
Yeah move cost to a new type and provide proper access through public method is a good design

Reply:
Looks like after drop cell on ProfileData, there's no need for a GasProfile (ProfileDataResult).